@types/react-native fix: <Animated.X> function prop args should infer original types

zanona opened the pull request:
following up https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41939#issuecomment-661022547 cc/ @alloy 

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

alloy commented on types/react-native/test/animated.tsx:
```diff
@@ -164,6 +164,27 @@ function TestAnimatedAPI() {
+            <Animated.View
+              style={{opacity: v1}}
+              //should infer (event: LayoutChangeEvent) => void;
+              onLayout={e => console.log(e.nativeEvent.layout)}
```
Because we specifically want to know that these args are not `any`, it might be better to expect an exact type, e.g.:

```suggestion
              onLayout={e => {
                e; // $ExpectType (event: LayoutChangeEvent) => void
              }}
```

That is, if that works at all; I don’t think I’ve tried it in a callback function before.

zanona commented:
Adding `$ExpectType` assertions seems to throw errors related the `opacity` as an animated value, which seems unrelated.

> ERROR: 170:15  expect  Compile error in typescript@3.7 but not in typescript@3.8.
Fix with a comment '// Minimum TypeScript Version: 3.8' just under the header.
Types of property 'opacity' are incompatible. Type 'Value' is not assignable to type 'number'.

Any ideas?

alloy commented:
> Adding `$ExpectType` assertions seems to throw errors related the `opacity` as an animated value, which looks unrelated.

That’s weird, so this is about the `opacity` key used in the `style` prop?

Perhaps moving the assertion into the callback function (see my original suggestion) side-steps checking of other props? Or casting the `v1` value as `any`?

> Tests are passing locally though.

How are you running the tests locally? (To ensure it also tests with TS < 3.8.)

zanona commented:
Thanks for your feedback, @alloy.
Apparently  the `opacity` error was some sort of glitch caused while asserting `<Animated.SectionList>`.

Basically, under `SectionList`, `renderItem` expects:

-  `(info: SectionListRenderItemInfo<number>)` on Typescript@4.0, but
- `(info:SectionListRenderItemInfo<any>)` on Typescript@3.5

However, this is also true for the original `<SectionList>`, in which case, I believe this may not be something we should be looking at here?

I added some notes to that test in particular, just so developers can understand why that it's not being tested at the moment.

I have also followed your suggestions and moved the assertions within the callback, so we only assert the argument types instead of the entire callback itself. Although, both declaration styles, would've behaved the same way.

Lastly, I have also added the `testID` props to each component just to identify each test, in case it's helpful.

Let me know in case I'm missing something here.

alloy commented:
This seems fine, thanks!
